### DISCUSSION TITLE
API to show data on dashboard

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#95] [Anything Else]
"As we currently use a dashboard that pulls in data via an API (or other connection methods) it would nice to be able to gave an API that can pull in details. E.g. /api/pageviews/last7days to get the last 7 days page views.

This would also be useful to make rankings of most visited posts on a blog.

Which pages get the most traffic. Ideally the API would be able to return specific page types, ie tell us the most popular product pages, ignoring home page, informational and support pages, for a specified time period."

**Update:** We now have the [stats API](https://plausible.io/docs/stats-api)

**Comment by binaryfire** [comment]
Awesome to see this is on the roadmap! It'd be good to have endpoints which can query data for specific URLs too.

**Comment by fabio-smu** [comment]
@metmarkosaric this issue is not on roadmap anymore?

**Comment by metmarkosaric** [comment]
thanks @fabio-smu! we've basically reduced our roadmap to only show the most immediate things we're working on. there's only a couple of items outstanding on the short-term roadmap. you can see it at https://github.com/plausible/analytics/projects/1. 

after that is completed, the API and the export/import of data will be the next two big projects we will focus on. both of these we want to do and they are some of the most upvoted features. we'll update the roadmap again as we complete the current tasks and as we prioritize the upcoming work. it is difficult to promise any timelines at this stage on when the API will be completed but we will keep you posted!

**Comment by alexphelps** [comment]
I'm also very interested in this, with an API to pull data for a dashboard, Plausible would be a great fit for us.

**Comment by lmiller1990** [comment]
Going to keep an eye on this issue - pretty keen. Will try to help out with coding if/when needed. The real value of Plausible, to me, is not plausible the hosted UI, but my analytics tracked in a non-invasive way. Access to that data is the next most important thing, this would let me perform my own analysis more easily.

Loving Plausible so far. 👍

**Comment by mishushakov** [comment]
Agreeing with all the things said above
our scenario: provide analytics for websites we host for our customers

having API would allow us to add the said websites to Plausible (automatically) and make it possible to fetch and display the data in our customer-facing console

**Comment by metmarkosaric** [comment]
We plan to work on the API as the next big project at the start of next year. In order to deliver the product that satisfies your needs, we would love to hear from people who want to offer Plausible Analytics to their customers, clients etc using this API.

What do you plan do to? What do you need and expect from the API? Do you have any specific requirements that we should consider when building this? 

You can share your feedback here or reach out to us on hello@plausible.io. Thanks!

**Comment by grstreten** [comment]
Great to hear that this is something actively on the horizon!

We love using the [Numerics](https://cynapse.com/numerics/) App, so an ability to integrate with this would be super slick.

Overall hopes would be at the least some historical reporting (maybe views over 7 days) and active right now.

Thanks for making such an awesome product ⚡️

**Comment by lmiller1990** [comment]
My use case for the API: I have a personal dashboard for my various socials, I want to add a small widget for my plausible data. I like to have everything in one place.

**Comment by Uninen** [comment]
This was the first thing I searched for after installing Plausible for a test on my sites.

Specifically what I'm looking for is an easy way to query statistics for a specific URL. I'm building a live streaming site for musicians, and each show has a unique URL. I'd like to offer the artists statistics about visits to their profile pages + visits to their live shows. It doesn't need to be realtime, a daily update would be enough for the first revision.

**Comment by janejuenyang** [comment]
Exciting stuff! In general, we’re looking for the ability to pull specific queries that aren’t necessarily highlighted or organized in the out-of-the-box Plausible dashboard. For instance, focusing on specific marketing campaigns we run and summarizing those conversion data in a single view. In another example, for A/B testing, we’d want to validate that the visitors that saw each variant are statistically the same to each other, which requires pulling data aggregated by that split and across different dimensions and running statistical tests.

**Reply by ukutaht** [reply]
Cheers @janejuenyang. I'm working on the API now so looking through these messages.

The way I'm approaching it is almost like a DB interface. The API will be very flexible and allow arbitrary ad-hoc queries so I think it will allow new interesting ways to analyze your stats.

I think to be able to do proper A/B testing we need to introduce an easy way to set session-level custom properties. This way you could segment visitors by which variant they saw. With the current capabilities you'd have to trigger a custom event for `Saw experiment: {variant: 'A'}` or similar. It would work but obviously not as streamlined as I'd like it to be.

**Comment by boberbober** [comment]
In the past year I've integrated Google Analytics, Search Console, Twitter, Facebook, Stripe, etc. to my live dashboard app and they all use OAuth 2.0 to generate access tokens for the API (except Twitter, they're on 1.0). It's pretty standard and makes it easier to share the access with a third party app. It would be great if Plausible followed that standard.

As for the API itself a simple GET endpoint that returns all the stats Plausible has to offer in JSON, configured by query parameters, is enough. No webhooks, account management, GraphQL support needed.

Looking forward to it!

**Comment by alexphelps** [comment]
We want to use Plausible in conjunction with our custom sites platform. We only need to create new "sites" via API and then basic methods to query the metrics by date range so that we can show the data in the customer dashboard. 

Really excited about this!

**Comment by luctus** [comment]
I'd love to integrate dailytics to Plausible, so as @boberbober mentioned, the most important thing is OAuth and a couple of GET endpoints to get data filtered by some params (url, start_date, end_date, etc)

Thanks in advance!

**Comment by evanlegamer** [comment]
We want to use Plausible with a fleet of website and get data for each website with GET request, but as we will use a dashboard for all our users , access and private key could be enough instead of Oauth.

**Reply by ukutaht** [reply]
Thanks for this. What sort of data would you like to get out of the API?

**Reply by evanlegamer** [reply]
We need to have a JSON answer, using parameters we could choose what we want to get from the API,  number of total visitors, number per pages, referral, device, etc... All we can have from the dashboard currently but accessible through an easy API. I think Oauth will not be a good choice for our solutions as everything is manage from our backend, and so we would have to use a service account for that and more... , i think a key could be enougth but proposing both for others can be a good solution too. Thanks for reading :)

**Reply by olafghanizadeh** [reply]
API Key authentication would be preferable for read only access. Getting data from GA through API is a bit cumbersome with oAuth.

**Comment by linuz90** [comment]
Excited about this! Regarding my use case, I'd love to be able to get the live visitors counter from Plausible to show it on a React frontend.

**Comment by olafghanizadeh** [comment]
JSON of visitor statistics would be nice, especially on a per page/url basis.

**Comment by ukutaht** [comment]
📢 I've been working on the early beta for this and it's ready for testing on plausible.io.

Full documentation can be found [here](https://plausible.io/docs/stats-api). Would love for everyone who's interested to test it out.

Please post all your questions and feedback as a reply to this message. If you think documentation could be improved, please let me know as well. It's still very early and we can make changes before releasing v1 🚧

**Reply by luctus** [reply]
It looks promising! I wonder if you are considering an endpoint to get page views per path (ordered by page views) and filtering by regex too.

Also, could I share my access token with a third party app?

Thanks in advance!

**Reply by ukutaht** [reply]
Thanks @luctus.

Yes I will work on `/api/v1/stats/breakdown` endpoint next week. It will allow to get top pages by pathname but also top sources, countries, etc.

Filtering by regex is not supported at the moment. We'll implement it later in the UI and API simultaneously.

> Also, could I share my access token with a third party app?

Yes, absolutely but you're responsible for keeping it secure :)

**Reply by luctus** [reply]
Amazing to hear! Thanks for the hard work!

**Reply by luctus** [reply]
I'm building a small API wrapper for Ruby here: 

https://github.com/dailytics/plausible-api [WIP]

**Reply by ukutaht** [reply]
Awesome @luctus. Any feedback on the API itself so far?

**Reply by luctus** [reply]
it works excellent! It wasn't that clear that by `site_id` you meant `domain`... apparently there is no reference to the concept  `site_id` in the UI ;)

I believe the only missing features are:
- being able to filter by regex
- get most visited content (based on that regex)

An endpoint that lists all your sites could be very handy too (thinking in the integration I want to do: simply add your token here and then continue with the wizard)

**Reply by luctus** [reply]
Hi @ukutaht! I have some doubts now...

1) I called to this endpoint `https://plausible.io/api/v1/stats/timeseries?site_id=dailytics.com&period=2d`
And I got 31 records (from 2021-01-11 to 2021-02-10... I was expecting to get just 2 records (for that `period=2d`)

2) I called to this endpoint `https://plausible.io/api/v1/stats/aggregate?site_id=dailytics.com&period=3mo&metrics=visitors,pageviews,bounce_rate,visit_duration` and then this one `https://plausible.io/api/v1/stats/aggregate?site_id=dailytics.com&period=3mo&metrics=visitors,pageviews,bounce_rate,visit_duration&filter=event%3Apage%3D%3D%2Ffeatures%2F` (I added the filter by `event:page`) and now I get more page views than the first request (with no filter)

3) I called to this endpoint `https://plausible.io/api/v1/realtime/visitors?site_id=dailytics.com` and I got an error (Oops! There's nothing here)

Once again, thanks a lot for this work! I'm probably doing something wrong here.

**Reply by ukutaht** [reply]
Thank you very much for the feedback @luctus. I will take a look today

**Reply by ukutaht** [reply]
Thanks I will make sure to note the site_id <-> domain equivalence in the docs.

### 1

The `2d` period is not recognized at the moment. Currently the only options are the same as in the UI: `day`, `7d`, `30d`. When we don't recognize the period, it defaults to `30d` which is is why you're seeing 31 records (30 historical days plus today). I can see why returning 31 days for `30d` can be confusing. And we probably should throw an error if you specify an unrecognized `period` value.

### 2

Sorry the docs are wrong in the 'Filtering' section. The parameter is `filters` not `filter`. Try again with that.

I think you're seeing more pageviews with each call because the API is real-time and includes all the new pageviews/visitors between the two API calls if the time range includes today.

In this case you're actually seeing stats for `30d` again because `3mo` is not recognised. Would make sense to support any number of days and months in the period parameter. I'll see what I can do.

### 3

Again my bad on documentation. The endpoint is `/api/v1/stats/realtime/visitors`. Notice the `stats`. I'll fix the docs.

**Reply by luctus** [reply]
Brilliant! Thanks a lot for the explanation @ukutaht! One last question... what is the expected format for the optional `date` paramater?

**Reply by ukutaht** [reply]
@luctus The format is ISO-8601 date format. I'm working on validating params at the moment, will add it to the docs and error messages in case of invalid format.

I also realized the docs are wrong in that the `date` param is not currently respected on `30d` and `7d` periods. I will fix that by v1.

**Comment by evanlegamer** [comment]
is there currently a way to add a domain to plausible by API?

**Reply by kurtmc** [reply]
It doesn't look like that has been done yet. I am currently using a work around in my terraform provider for plausible. I have written some client code that can create/update/delete sites if you are interested: https://github.com/mcalpinefree/terraform-provider-plausible/blob/master/plausibleclient/site.go#L12

I would be happy to split my plausible client to a new repo if anyone is interested, and of course I will update my client to use the proper API once the features I need are there.

**Reply by metmarkosaric** [reply]
This is only the first version of our API which is read only with the purpose of taking the stats and displaying them in other places. We plan to work on adding more functions later. You can follow our plans for the API that allows you to add sites here: https://github.com/plausible/analytics/discussions/150

**Comment by mishushakov** [comment]
would be awesome if we could access plausible on behalf of our customers
they'd give us their Plausible API credentials and we in turn would send events/page-views to their dashboards

would be an amazing feature to offer from our side for all the nice people already using plausible!

i'm already very grateful for the current work you've done so-far on the API
many thanks

**Comment by stevenferrer** [comment]
Hi, thank you for the amazing work!

I was just wondering if there would be support for complex filters?  For example, I have a URL path `/products/my-product-1234` and I want to query the number of views for this particular product. 

Currently, I am getting it directly from ClickHouse using the query below.

```sql
select uniq(e.user_id) unique_views,
	count(*) total_views
from plausible_events_db.events e
join (
	select session_id 
	from plausible_events_db.sessions s 
	where s."domain" = 'my-domain.com' 
) s
on e.session_id = s.session_id
where e."domain" = 'my-domain.com' and
	e.name = 'pageview' and	
	e.pathname like '%/products/%' and
	splitByString('-', e.pathname)[-1] = 'product-id'
```

**Reply by ukutaht** [reply]
Hey, the API supports filtering on a pathname. I consider it a simple filter since it only supports equality and filters can only be combined with AND (no support for OR).

Here's how you would get the same data through the API: https://plausible.io/docs/stats-api#number-of-visitors-to-a-specific-page

The API is live for testing on our plausible.io URL. For self-hosted installations, you can update from the `master` tag but do keep in mind this is an unstable tag with no guarantees.

**Reply by stevenferrer** [reply]
I just realized that I was only using the query to get the views by _product ID_ instead of the full URL path. :facepalm: 

Many thanks for the quick response!

**Comment by ukutaht** [comment]
On Monday I will be merging #723 with some updates:
* Support for `custom` date range
* Support for `/api/v1/stats/breakdown` endpoint.
* ⚠️⚠️⚠️ Breaking change to `/api/v1/stats/timeseries` response format ⚠️⚠️⚠️

I will update the docs on Monday but here's a heads up on the breaking change.

The response format for timeseries will have two changes:
1. The list is now wrapped in an object for future flexibility
2. Rename `value` -> `visitors` so we could return different metrics in the future

Old format:
```json
[
    {
        "date": "2020-08-01",
        "value": 36085
    },
    {
        "date": "2020-09-01",
        "value": 27688
    },
    {
        "date": "2020-10-01",
        "value": 71615
    },
    {
        "date": "2020-11-01",
        "value": 31440
    },
    {
        "date": "2020-12-01",
        "value": 35804
    },
    {
        "date": "2021-01-01",
        "value": 0
    }
]
```

New format:
```json
{
  "results":  [
    {
        "date": "2020-08-01",
        "visitors": 36085
    },
    {
        "date": "2020-09-01",
        "visitors": 27688
    },
    {
        "date": "2020-10-01",
        "visitors": 71615
    },
    {
        "date": "2020-11-01",
        "visitors": 31440
    },
    {
        "date": "2020-12-01",
        "visitors": 35804
    },
    {
        "date": "2021-01-01",
        "visitors": 0
    }
  ]
}
```

**Reply by luctus** [reply]
Thanks a lot for that! I'll start updating the ruby gem then ;)

**Reply by ukutaht** [reply]
Cheers, this change is now live

**Reply by luctus** [reply]
Amazing! I plan to finish the gem update tonight.

**Reply by luctus** [reply]
The API worked like a charm! Just in case you want to take a look at some Ruby code, here is the gem: https://github.com/dailytics/plausible_api

**Reply by ukutaht** [reply]
Sweet, thanks @luctus !

**Comment by Nick-Mazuk** [comment]
This looks great!

If you're still building out API features, I'd find a lot of use being able to query by specific events. Right now, it looks like you can query by page views or page view statistics and filter by events, but you can't query the events themselves (could be wrong).

For instance, let's say you have a blog and want to know the CTR for the previews. You could setup two events: a) when the user is shown the preview for a blog, and 2) when the user clicks on a preview. Now this could happen on any page, and the user may click on the same preview multiple times on that page. So there's no good way to query a page view then filter it for this event. I would find being able to query by events extremely useful.

**Reply by ukutaht** [reply]
Thanks!

I will add support for events by the time this is stable v1. Hopefully I can work on this later this week or some time next week.

**Reply by Nick-Mazuk** [reply]
Great, thanks @ukutaht!

**Comment by shornuk** [comment]
This has been great to start playing with. I've created a plugin to pull data through to a Craft cms installation. (https://plugins.craftcms.com/plausible)

I'm just working on some updates and adding more features, but I've hit a couple of errors/bugs - where's the best place to post these? I'm aware it's beta, but guessing you'd rather know about these this. 

For instance, I've using the aggregate endpoint to get a summary/overview. I've added in the `compare` parameter, which works fine, but when I choose 6mo or 12mo as the period, it throws an error  `(ArithmeticError) bad argument in arithmetic expression`.
Every other time period works fine.

**Reply by ukutaht** [reply]
@shornuk Thanks for letting me know. And yes, this is exactly the sort of feedback I was hoping from the beta. Could you open a single issue and list all problems there? We track bugs in the issue section.

**Reply by shornuk** [reply]
Great, will do. 👍

**Comment by evanlegamer** [comment]
Hi guys,
just a question, how do i enable the api on a self hosted solution ?
Thank you in advance.

**Reply by Nick-Mazuk** [reply]
My guess is that right now, you can't.

By looking at their blog and comparison between the cloud and self-hosted, only stable features are added to the self-hosted option. Since the API is in development, it's probably not reached "stable" yet.

https://plausible.io/blog/building-open-source#self-hosted-analytics-will-become-a-twice-a-year-long-term-release
https://plausible.io/self-hosted-web-analytics

**Reply by metmarkosaric** [reply]
thank you @Nick-Mazuk!

**Reply by j4w8n** [reply]
The API is now working on self-hosted - if you don't mind installing a release candidate. Change your docker-compose.yml file to install `v1.3.0-rc.1`

```
...

plausible:
  image: plausible/analytics:v1.3.0-rc.1

...
```

```
docker-compose down
docker-compose pull plausible
docker-compose up -d
```

**Comment by Jamesking56** [comment]
I run a web development agency - I am in the process of creating a custom client area for my clients that integrates all of the services and websites we manage for them. I would love to include Plausible as an analytics offering and show the plausible stats inside of the client area.

This way, the client only needs 1 login and to see everything about their website. We can then manage a single Plausible account for all of our clients.

**Reply by metmarkosaric** [reply]
thanks @Jamesking56! this is possible already. our stats API is available and ready to go. see https://plausible.io/docs/stats-api

**Reply by Jamesking56** [reply]
Cool, I guess the only thing we're missing is the ability for our client area to be able to add / remove sites in Plausible

**Reply by metmarkosaric** [reply]
yeah, at the moment you'll need to manually create the sites. we plan to allow this to be done using the api in the future. #150

**Comment by ukutaht** [comment]
I'm just finishing work on hopefully the last round of improvements before calling the API stable.

Here's the full PR: https://github.com/plausible/analytics/pull/826. I will deploy the changes to our live environment tomorrow at noon UTC. If no bug reports come up during next week then I think it's ready to go.

Hightlights:

* :warning: BREAKING: Wrap results of the aggregate endpoint in {"results": ..} for consistency with other endpoints
* :warning: BREAKING: Show correct number of days in 7d and 30d periods (currently it's off by one, 8 days returned for 7 day period etc)
* :warning: BREAKING: Return '(Direct / None)' instead of empty string for missing source, referrer, utm params
* Implement timeseries endpoint for a single day with hourly interval. Fixes #797
* Fix aggregate endpoint comparison with previous period.  Fixes #792
* Fix/Align breakdown query with current queries in the UI. Fixes #829
* Add event name as a property for filtering and breakdowns
* Add custom props as a property for filtering and breakdowns

**Reply by shornuk** [reply]
Awesome. Got a load of improvements to my CraftCMS plugin that I can test with tomorrow 👍

**Reply by luctus** [reply]
Amazing! I'll be updating the ruby gem as soon as I can ;)

**Reply by shornuk** [reply]
Just wondering if these changes are live? I'm still running into the same issues/errors.

**Reply by ukutaht** [reply]
Sorry you're right, I missed Friday noon. I've deployed the changes now.

**Reply by luctus** [reply]
💎   Ruby gem updated! 💎

**Reply by alexphelps** [reply]
Will it be eventually possible to add websites and delete websites via API?

**Reply by metmarkosaric** [reply]
hi @alexphelps! yes, shouldn't take too long before the first version of sites API is out. it allows you to add sites and we're testing it. you can follow the progress here https://github.com/plausible/analytics/discussions/150

**Comment by metmarkosaric** [comment]
The site provisioning API is now in private preview. It's a work in progress but currently it allows 2 basic operations:

* Create a new site
* Find or create a shared link by name

You can learn more here https://plausible.io/docs/sites-api

**Reply by binaryfire** [reply]
@metmarkosaric Nice work! Do you guys have an ETA on when it'll hit the self-hosted edition?

Regarding other endpoints, I think your suggestions are great. The crucial ones for me are deleting and updating sites. But configuring spike notifications and custom domains would be nice to have too.

**Reply by binaryfire** [reply]
@metmarkosaric Hey Marko, just checking in again. Will this be available in the open source version soon?

**Reply by metmarkosaric** [reply]
Sites API is being tested in private beta. Regularly features are battle-tested in the cloud which allows us to fix any bugs before the general self-hosted release. Every six months or so we combine all the changes into a new self-hosted release. Last self-hosted release was on April 14th 2021 so I would think the next one should happen at some point before end of this year. Not sure yet if the sites API will make it to the next self-hosted release or not as it's not even open to the public on the cloud yet.

**Comment by veterinarymoda** [comment]
Hi
We'd like some help with using the Plausible API to consolidate a couple of data sets into a single online report that is also exportable - pageviews, button clicks - nothing too complicated - happy to pay a modest fee

**Comment by SimonKlausLudwig** [comment]
Is there any plan to support graphql or Swagger ?

**Reply by metmarkosaric** [reply]
There are no plans from our side. We're a small team with limited development resources so need to focus on more frequently asked for requests. Our [stats API](https://plausible.io/docs/stats-api) allows you to take any data that you want and then you can bring it wherever you want using different tools. Perhaps someone will be able to build connectors from our API to other tools in the future. In that case, we're happy to list them in our docs alongside other community built integrations

**Comment by evanlegamer** [comment]
Hi guys, Do the Site provisioning API reference is available in the 1.4.0 ?

**Reply by ukutaht** [reply]
Yes it is

**Comment by RobertJoonas** [comment]
Our site Provisioning API now also supports deleting existing sites from Plausible. Take a look here for the documentation: https://plausible.io/docs/sites-api#delete-apiv1sitessite_id.

**Reply by evanlegamer** [reply]
is it available in the 1.4.0 version ?

**Reply by ukutaht** [reply]
No. As you can see in the [Changelog](https://github.com/plausible/analytics/blob/master/CHANGELOG.md), site deletion was added after 1.4 release
